graph/push.go

```diff
@@ -287,63 +311,66 @@ func (s *TagStore) pushV2Repository(r *registry.Session, eng *engine.Engine, out
 		if err != nil {
 			return err
 		}
-		log.Infof("Signed manifest using daemon's key: %s", s.trustKey.KeyID())
-
-		manifestBytes = string(signedBody)
-	}
-
-	manifest, verified, err := s.verifyManifest(eng, []byte(manifestBytes))
-	if err != nil {
-		return fmt.Errorf("error verifying manifest: %s", err)
-	}
-
-	if err := checkValidManifest(manifest); err != nil {
-		return fmt.Errorf("invalid manifest: %s", err)
-	}
-
-	if !verified {
-		log.Debugf("Pushing unverified image")
-	}
+		log.Infof("Signed manifest for %s:%s using daemon's key: %s", repoInfo.LocalName, tag, s.trustKey.KeyID())
 
-	for i := len(manifest.FSLayers) - 1; i >= 0; i-- {
-		var (
-			sumStr  = manifest.FSLayers[i].BlobSum
-			imgJSON = []byte(manifest.History[i].V1Compatibility)
-		)
+		manifestBytes := string(signedBody)
 
-		sumParts := strings.SplitN(sumStr, ":", 2)
-		if len(sumParts) < 2 {
-			return fmt.Errorf("Invalid checksum: %s", sumStr)
+		manifest, verified, err := s.verifyManifest(eng, signedBody)
+		if err != nil {
+			return fmt.Errorf("error verifying manifest: %s", err)
```

Comment:
I understand this is confusing mostly because there are two types of verification, signature verification and key verification.  An error will be throw here if the signature is invalid, which there is no case in which we should accept a manifest with a signature that failed.  This would mean that the contents have changed in some way. The verification boolean refers to whether or not the key has been verified as able to build an image with the given name.  Both are required for a fully secure install, however a key verification failure is recoverable and should be treated as built by an untrusted source. This verification check is mostly to produce the log line about whether or not the build is verifiable. Currently it does not have much value outside of trusted official builds.

The consequence of throwing an error here is a failure to push, but this error would indicate something really wrong happened since the signature was just generated.